File: java/src/main/java/org/example/crypto/Pbkdf2Example.java
package org.example.crypto;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;

public final class Pbkdf2Example {
    public record DerivedKey(byte[] salt, int iterations, byte[] key) {
        public DerivedKey {
            salt = salt.clone();
            key = key.clone();
        }
        @Override public byte[] salt() { return salt.clone(); }
        @Override public byte[] key() { return key.clone(); }
    }

    private Pbkdf2Example() {}

    public static DerivedKey derive(char[] password) throws Exception {
        byte[] salt = CryptoSupport.randomBytes(16);
        int iterations = 600_000; // Calibrate for the deployment environment.
        PBEKeySpec spec = new PBEKeySpec(password, salt, iterations, 256);
        try {
            byte[] key = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
                    .generateSecret(spec).getEncoded();
            return new DerivedKey(salt, iterations, key);
        } finally {
            spec.clearPassword();
        }
    }

    public static boolean verify(char[] password, DerivedKey stored) throws Exception {
        PBEKeySpec spec = new PBEKeySpec(password, stored.salt(), stored.iterations(), 256);
        try {
            byte[] candidate = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
                    .generateSecret(spec).getEncoded();
            try { return CryptoSupport.constantTimeEquals(candidate, stored.key()); }
            finally { Arrays.fill(candidate, (byte) 0); }
        } finally {
            spec.clearPassword();
        }
    }

    public static void main(String[] args) throws Exception {
        char[] password = "correct horse battery staple".toCharArray();
        try {
            DerivedKey stored = derive(password);
            CryptoSupport.require(verify(password, stored), "PBKDF2 verification failed");
            System.out.println("Pbkdf2Example salt=" + CryptoSupport.hex(stored.salt()));
        } finally {
            Arrays.fill(password, '\0');
        }
    }
}

File: python/pbkdf2_example.py
"""PBKDF2-HMAC-SHA-256 password verification with stored salt and work factor."""

from dataclasses import dataclass
from cryptography.exceptions import InvalidKey
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

from crypto_support import random_bytes

ITERATIONS = 600_000  # Calibrate for the deployment environment.


@dataclass(frozen=True)
class StoredPassword:
    salt: bytes
    iterations: int
    derived_key: bytes


def _kdf(salt: bytes, iterations: int) -> PBKDF2HMAC:
    return PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=iterations,
    )


def derive(password: bytes) -> StoredPassword:
    salt = random_bytes(16)
    return StoredPassword(salt, ITERATIONS, _kdf(salt, ITERATIONS).derive(password))


def verify(password: bytes, stored: StoredPassword) -> bool:
    try:
        _kdf(stored.salt, stored.iterations).verify(password, stored.derived_key)
        return True
    except InvalidKey:
        return False


def main() -> None:
    stored = derive(b"correct horse battery staple")
    assert verify(b"correct horse battery staple", stored)
    assert not verify(b"incorrect", stored)
    print(f"Pbkdf2Example salt={stored.salt.hex()} iterations={stored.iterations}")


if __name__ == "__main__":
    main()
